src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -273,102 +224,443 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
+            # loop iterations apart the same memory location will be accessed).
+            # If this should be 0, it means no dependency. Though even here
+            # loop boundaries could be used for further checks - e.g. if
+            # the distance is N with N loop iterations, e.g.:
+            # do i=1, N: a(i)=a(i+N)
+            # Then there would still be no dependency.
+            return sol
+
+        return None
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def independent_multi_subscript(var_name, write_access, other_access,
+                                    subscripts):
+        '''Test multiple subscripts that share variables. This includes cases
+        like `a(i,i) = a(i,i+1)` or `a(i, indx(i)) = a(i,5)` etc.
+        At this stage only a minimal test is done: if there is one subscript
+        that is independent, the whole access can be parallelised, e.g.
+        `a(i,i+j) = a(i,j*j-i*i)` if the 'i' loop is parallelised. Even though
+        the second subscript might be identical for different i and j values,
+        the fact that the first subscript is independent makes the access
+        parallelisable.
+
+        :param str var_name: the name of the loop variable of the loop to be \
+            parallelised.
+        :param write_access: access information a single write access.
+        :type write_access: :py:class:`psyclone.core.access_info.AccessInfo`
+        :param other_access: access information the other (read or write) \
+            access.
+        :type other_access: :py:class:`psyclone.core.access_info.AccessInfo`
+
+        :returns: whether the two accesses can be parallelised or not.
+        :type: bool
+
+        '''
+        # If we find one subscript that is independent, the loop can be
+        # parallelised. E.g. `a(i, index(i)) = a(i, 5)`. The fact that
+        # the first subscript is i, means that each different iteration
+        # will access a different column, even if index(i) is 5.
+
+        for ind in subscripts:
+            index_written = write_access.component_indices[ind]
+            index_other = other_access.component_indices[ind]
+            distance = DependencyTools.get_dependency_distance(var_name,
+                                                               index_written,
+                                                               index_other)
+            if distance == 0:
+                # Notice that distance 0 will only be returned if the
+                # subscript actually depends on the loop variable.
+                return True
+        # Additional tests could be added here (e.g. to see that
+        # `a(i,i)=a(i,i+1)+1` can be parallelised).
+
+        return False
+
+    # -------------------------------------------------------------------------
+    def is_array_access_parallelisable(self, loop_variables, write_access,
```

Comment:
You can ignore this suggestion but maybe you could name this method: `is_loop_carried_dependency` since this seems to match the docstring and it will make it more differentiable from the next method (then also the returns docstring).